presto-tests/src/test/java/com/facebook/presto/server/TestQueryResource.java

```diff
@@ -131,6 +131,67 @@ public void testGetQueryInfos()
         assertStateCounts(infos, 0, 5, 0, 0);
     }
 
+    @Test
+    public void testGuavaTryAcquireShouldReturnImmediatelyWithoutToken()
+    {
+        RateLimiter rateLimiter = RateLimiter.create(1);
+        rateLimiter.acquire();
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        boolean result = rateLimiter.tryAcquire(3);
+        stopwatch.stop();
+        long elapsed = stopwatch.elapsed(MILLISECONDS);
+        assertFalse(result);
+        assertEquals(elapsed, 0);
+    }
+
+    @Test
+    public void testBlockingRateLimitShouldNotDelay()
+            throws Exception
+    {
+        runner = createQueryRunner(ImmutableMap.of("query.client.timeout", "10s", "query-manager.rate-limit-bucket-max-size", "100"));
+        server = runner.getCoordinator();
+        long maxRetry = 15;
+        final URI sameURI = getQueuedURI("SELECT 1");
+        int i = 0;
+        Stopwatch stopwatch = Stopwatch.createStarted();
+        while (i < maxRetry) {
+            Request request = prepareGet()
+                    .setHeader(PRESTO_USER, "user")
+                    .setUri(sameURI)
+                    .build();
+            client.execute(request, createJsonResponseHandler(jsonCodec(QueryResults.class)));
+            i++;
+        }
+        stopwatch.stop();
+        long millis = stopwatch.elapsed(MILLISECONDS);
+        //Should not be rate limited when allowing 100/s, using 6 seconds to avoid flakiness
+        assertTrue(millis < 6000);
+    }
+
+    @Test
+    public void testBlockingRateLimitShouldDelay()
+            throws Exception
+    {
+        runner = createQueryRunner(ImmutableMap.of("query.client.timeout", "10s", "query-manager.rate-limit-bucket-max-size", "1"));
+        server = runner.getCoordinator();
+        long maxRetry = 15;
+        final URI sameURI = getQueuedURI("SELECT 1");
+        int i = 0;
+        Stopwatch stopwatch = Stopwatch.createStarted();
```

Comment:
Looks like the logic structure of those two functions are nearly the same, can we extract the common logic and avoid duplicate logic